compute/metadata/metadata_test.go

```diff
@@ -44,15 +45,55 @@ func TestOnGCE_Stress(t *testing.T) {
 }
 
 func TestOnGCE_Force(t *testing.T) {
+	ctx := context.Background()
 	onGCEOnce = sync.Once{}
 	old := os.Getenv(metadataHostEnv)
 	defer os.Setenv(metadataHostEnv, old)
 	os.Setenv(metadataHostEnv, "127.0.0.1")
-	if !OnGCE() {
+	if !OnGCEWithContext(ctx) {
 		t.Error("OnGCE() = false; want true")
 	}
 }
 
+func TestOnGCE_Cancel(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	onGCEOnce = sync.Once{}
+	if OnGCEWithContext(ctx) {
+		t.Error("OnGCE() = true; want false")
+	}
+}
+
+func TestOnGCE_CancelTryHarder(t *testing.T) {
+	// If system info suggests GCE, we allow extra time for the
+	// probe with higher latency (HTTP or DNS) to return. In this
+	// test, the DNS probe fails organically, but we spoof the
+	// system info suggestion so that it waits longer for the HTTP
+	// probe to return. However, that additional wait budget should
+	// still be controlled by the calling context.
+	defaultSystemInfoSuggestsGCE = true
+	defer func() {
+		defaultSystemInfoSuggestsGCE = false
+	}()
```

Comment:
this is pretty kludgy and could cause a data race if there are multiple tests running in parallel. in practice i don't think this a concern because we aren't using `t.Parallel` in these package tests.

Reply:
done!